File: tests/algorithms/x/Java/dynamic_programming/max_subarray_sum.java
public class Main {
    static double[] empty = new double[0];

    static double max_subarray_sum(double[] nums, boolean allow_empty) {
        if (nums.length == 0) {
            return 0.0;
        }
        double max_sum = 0.0;
        double curr_sum = 0.0;
        if (((Boolean)(allow_empty))) {
            max_sum = 0.0;
            curr_sum = 0.0;
            int i = 0;
            while (i < nums.length) {
                double num = nums[i];
                double temp = curr_sum + num;
                curr_sum = temp > 0.0 ? temp : 0.0;
                if (curr_sum > max_sum) {
                    max_sum = curr_sum;
                }
                i = i + 1;
            }
        } else {
            max_sum = nums[0];
            curr_sum = nums[0];
            int i_1 = 1;
            while (i_1 < nums.length) {
                double num_1 = nums[i_1];
                double temp_1 = curr_sum + num_1;
                curr_sum = temp_1 > num_1 ? temp_1 : num_1;
                if (curr_sum > max_sum) {
                    max_sum = curr_sum;
                }
                i_1 = i_1 + 1;
            }
        }
        return max_sum;
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{2.0, 8.0, 9.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{0.0, 0.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{-1.0, 0.0, 1.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{1.0, 2.0, 3.0, 4.0, -2.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{-2.0, 1.0, -3.0, 4.0, -1.0, 2.0, 1.0, -5.0, 4.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{2.0, 3.0, -9.0, 8.0, -2.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{-2.0, -3.0, -1.0, -4.0, -6.0})), false)));
            System.out.println(_p(max_subarray_sum(((double[])(new double[]{-2.0, -3.0, -1.0, -4.0, -6.0})), true)));
            empty = ((double[])(new double[]{}));
            System.out.println(_p(max_subarray_sum(((double[])(empty)), false)));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/dynamic_programming/max_subarray_sum.pas
{$mode objfpc}
program Main;
uses SysUtils, Math;
type RealArray = array of real;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  empty: array of real;
  allow_empty: boolean;
  nums: RealArray;
function max_subarray_sum(nums: RealArray; allow_empty: boolean): real; forward;
function max_subarray_sum(nums: RealArray; allow_empty: boolean): real;
var
  max_subarray_sum_max_sum: real;
  max_subarray_sum_curr_sum: real;
  max_subarray_sum_i: integer;
  max_subarray_sum_num: real;
  max_subarray_sum_temp: real;
begin
  if Length(nums) = 0 then begin
  exit(0);
end;
  max_subarray_sum_max_sum := 0;
  max_subarray_sum_curr_sum := 0;
  if allow_empty then begin
  max_subarray_sum_max_sum := 0;
  max_subarray_sum_curr_sum := 0;
  max_subarray_sum_i := 0;
  while max_subarray_sum_i < Length(nums) do begin
  max_subarray_sum_num := nums[max_subarray_sum_i];
  max_subarray_sum_temp := max_subarray_sum_curr_sum + max_subarray_sum_num;
  if max_subarray_sum_temp > 0 then begin
  max_subarray_sum_curr_sum := max_subarray_sum_temp;
end else begin
  max_subarray_sum_curr_sum := 0;
end;
  if max_subarray_sum_curr_sum > max_subarray_sum_max_sum then begin
  max_subarray_sum_max_sum := max_subarray_sum_curr_sum;
end;
  max_subarray_sum_i := max_subarray_sum_i + 1;
end;
end else begin
  max_subarray_sum_max_sum := nums[0];
  max_subarray_sum_curr_sum := nums[0];
  max_subarray_sum_i := 1;
  while max_subarray_sum_i < Length(nums) do begin
  max_subarray_sum_num := nums[max_subarray_sum_i];
  max_subarray_sum_temp := max_subarray_sum_curr_sum + max_subarray_sum_num;
  if max_subarray_sum_temp > max_subarray_sum_num then begin
  max_subarray_sum_curr_sum := max_subarray_sum_temp;
end else begin
  max_subarray_sum_curr_sum := max_subarray_sum_num;
end;
  if max_subarray_sum_curr_sum > max_subarray_sum_max_sum then begin
  max_subarray_sum_max_sum := max_subarray_sum_curr_sum;
end;
  max_subarray_sum_i := max_subarray_sum_i + 1;
end;
end;
  exit(max_subarray_sum_max_sum);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(FloatToStr(max_subarray_sum([2, 8, 9], false)));
  writeln(FloatToStr(max_subarray_sum([0, 0], false)));
  writeln(FloatToStr(max_subarray_sum([-1, 0, 1], false)));
  writeln(FloatToStr(max_subarray_sum([1, 2, 3, 4, -2], false)));
  writeln(FloatToStr(max_subarray_sum([-2, 1, -3, 4, -1, 2, 1, -5, 4], false)));
  writeln(FloatToStr(max_subarray_sum([2, 3, -9, 8, -2], false)));
  writeln(FloatToStr(max_subarray_sum([-2, -3, -1, -4, -6], false)));
  writeln(FloatToStr(max_subarray_sum([-2, -3, -1, -4, -6], true)));
  empty := [];
  writeln(FloatToStr(max_subarray_sum(empty, false)));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
